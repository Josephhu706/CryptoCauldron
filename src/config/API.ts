import axios from "axios";

export type Coin = {
    id: string;
    symbol: string;
    name: string;
    image: string;
    current_price: number;
    market_cap: number;
    market_cap_rank: number;
    fully_diluted_valuation: number;
    total_volume: number;
    high_24h: number;
    low_24h: number;
    price_change_24h: number;
    price_change_percentage_24h: number;
    market_cap_change_24h: number;
    market_cap_change_percentage_24h: number;
    circulating_supply: number;
    total_supply: number;
    max_supply: number;
    ath: number;
    ath_change_percentage: number;
    ath_date: string;
    atl: number;
    atl_change_percentage: number;
    atl_date: string;
    roi: any;
    last_updated: string;
}

export const CoinList = () =>
`https://api.coingecko.com/api/v3/coins/markets?vs_currency=aud&order=market_cap_desc&per_page=100&page=1&sparkline=false`;

export const SingleCoin = (id:number) =>
`https://api.coingecko.com/api/v3/coins/${id}`;

export const HistoricalChart = (id:number, days = 365) =>
`https://api.coingecko.com/api/v3/coins/${id}/market_chart?vs_currency=aud&days=${days}`;

export const TrendingCoins = () =>
`https://api.coingecko.com/api/v3/coins/markets?vs_currency=aud&order=gecko_desc&per_page=10&page=1&sparkline=false&price_change_percentage=24h`;

export const fetchCrypto = async ()=>{
    //put page not found 
    const endpoint = `https://api.coingecko.com/api/v3/coins/markets?vs_currency=aud&order=market_cap_desc&per_page=100&page=1&sparkline=false`;
    const {data} = await axios.get(endpoint)
    console.log(data)
    return data

}