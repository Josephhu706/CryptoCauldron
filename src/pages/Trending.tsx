import React, { useEffect, useState } from "react";
import { fetchCrypto, Coin } from "../config/API";
import { ThemeProvider, createTheme } from "@mui/material/styles";
import { motion } from "framer-motion";
import "../styles/TrendingStyles.scss";
import { CssBaseline } from "@mui/material/";
import Lottie from "react-lottie-player";
import animationData from "../assets/images/cauldron.json";
import CircularProgress from "@mui/material/CircularProgress";
import callError from "../assets/images/call-error.svg";
import Typography from "@mui/material/Typography";
import MotionWrap from "../wrapper/MotionWrap";
import { useNavigate } from "react-router-dom";

const darkTheme = createTheme({
  palette: {
    mode: "dark",
  },
});

const Trending = () => {
  const [trending, setTrending] = useState<Coin[]>([]);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<string>('');

  const navigate = useNavigate();

  useEffect(() => {
    getCoins();
  }, []);

  const getCoins = async () => {
    setLoading(true);
    try {
      const newCoins = await fetchCrypto();
      let sortCoins: Array<Coin> = newCoins;
      sortCoins.sort((a, b) => {
        return b.total_volume - a.total_volume;
      });
      setTrending(sortCoins.slice(0, 10));
    } catch (error: any) {
      setError(error);
    }
    setLoading(false);
  };

  return (
    <div data-testid="trendingPage">
      <Typography
      data-testid="title"
        className="trendingTitle"
        color="common.white"
        variant="h4"
        style={{ margin: 18, fontFamily: "Montserrat" }}
      >
        Top 10 Trending Coins By Volume
      </Typography>
      <div className="app__header app__flex" data-testid="TrendingPage">
        <ThemeProvider theme={darkTheme}>
          <CssBaseline />
          {loading ? (
            <CircularProgress data-testid="loading" />
          ) : (error ? 
            <div data-testid="fetchError">
              <img
                src={callError}
                style={{ height: "20em" }}
                alt="fetchError"
              />
            </div>
           : 
            <div className="circlesContainer">
              <motion.div
                initial={{ y: 300, opacity: 0 }}
                animate={{ y: -10, opacity: 1 }}
                transition={{ delay: 0.2, type: "tween", duration: 2 }}
                viewport={{ once: true }}
                className="app__header-circlesTop3"
              >
                <div className="circleContent">
                {trending.slice(0, 3).map((circle, index) => (
                  <motion.div
                  data-testid="coinCircle"
                    onClick={() => navigate(`/coins/${circle.id}`)}
                    whileInView={{ opacity: 1 }}
                    whileHover={{ scale: 1.2 }}
                    transition={{ duration: 0.5, type: "tween" }}
                    className="circle-cmp top3 app__flex"
                    key={`circle-${index}`}
                  >
                    <div className="overlay">
                      <h3 className="rank">{index + 1}</h3>
                    </div>
                    <img src={circle.image} alt="circle" />
                    <div className="coinName">{circle.name}</div>
                  </motion.div>
                ))}
                </div>
              </motion.div>
              <motion.div
                initial={{ y: 300, opacity: 0 }}
                animate={{ y: -10, opacity: 1 }}
                transition={{ delay: 0.3, type: "tween", duration: 2 }}
                viewport={{ once: true }}
                className="app__header-circlesTop7"
              >
                <div className="circleContent">
                {trending.slice(3, 7).map((circle, index) => (
                  <motion.div
                    data-testid="coinCircle"
                    onClick={() => navigate(`/coins/${circle.id}`)}
                    whileInView={{ opacity: 1 }}
                    whileHover={{ scale: 1.2 }}
                    transition={{ duration: 0.5, type: "tween" }}
                    className="circle-cmp app__flex"
                    key={`circle-${index}`}
                  >
                    <div className="overlay">
                      <h3 className="rank">{index + 4}</h3>
                    </div>
                    <img src={circle.image} alt="circle" />
                    <div className="coinName">{circle.name}</div>
                  </motion.div>
                ))}
                </div>
              </motion.div>
              <motion.div
                initial={{ y: 300, opacity: 0 }}
                animate={{ y: -10, opacity: 1 }}
                transition={{ delay: 0.6, type: "tween", duration: 2 }}
                viewport={{ once: true }}
                className="app__header-circlesTop10"
              >
                <div className="circleContent">
                {trending.slice(7).map((circle, index) => (
                  <motion.div
                    data-testid="coinCircle"
                    onClick={() => navigate(`/coins/${circle.id}`)}
                    whileInView={{ opacity: 1 }}
                    whileHover={{ scale: 1.2 }}
                    transition={{ duration: 0.5, type: "tween" }}
                    className="circle-cmp app__flex"
                    key={`circle-${index}`}
                  >
                    <div className="overlay">
                      <h3 className="rank">{index + 8}</h3>
                    </div>
                    <img src={circle.image} alt="circle" />
                    <div className="coinName">{circle.name}</div>
                  </motion.div>
                ))}
                </div>
              </motion.div>
            </div>
          )}
          <div data-testid="animation">
            <Lottie
              className="animation"
              loop
              animationData={animationData}
              play
              style={{ width: 720, height: 400 }}
            />
          </div>
        </ThemeProvider>
      </div>
    </div>
  );
};

export default MotionWrap(Trending);
