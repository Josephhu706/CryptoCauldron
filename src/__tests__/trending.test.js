import { render, screen, cleanup, waitForElement, getByTestId, fireEvent, waitFor, act} from '@testing-library/react'
import Trending from '../pages/Trending'
import '@testing-library/jest-dom'
import axiosMock from 'axios'
import renderer from 'react-test-renderer'
import 'jest-canvas-mock';
import {fetchCrypto} from '../config/API'

// //makes sure each test is starting from the same starting point
afterEach(()=>{
    cleanup()
})

jest.mock('axios');

it('should run', async () => {
    act(()=>render(<Trending />)) 
})
 

const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
   ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate,
}));


test("trending should be rendered", () => {
    render(<Trending />);
    const title = screen.getByTestId("title");
    expect(title).toBeInTheDocument()
});

test("title should be rendered", () => {
    render(<Trending />);
    const trendingPage = screen.getByTestId("trendingPage");
    expect(trendingPage).toBeInTheDocument()
});


test("cauldron should be rendered", () => {
    render(<Trending />);
    const cauldron = screen.getByTestId("animation");
    expect(cauldron).toBeInTheDocument()
});

test("Coins should be rendered after fetching and can navigate to coin page", async () => {
  axiosMock.get.mockResolvedValueOnce({data: [            
      {"id": "bitcoin",
      "symbol": "btc",
      "name": "Bitcoin",
      "image": "https://assets.coingecko.com/coins/images/1/large/bitcoin.png?1547033579",
      "current_price": 39161,
      "market_cap": 746975620874,
      "market_cap_rank": 1,
      "fully_diluted_valuation": 822780677096,
      "total_volume": 37007140684,
      "high_24h": 41702,
      "low_24h": 38967,
      "price_change_24h": -2356.366268313068,
      "price_change_percentage_24h": -5.67563,
      "market_cap_change_24h": -44140206844.06592,
      "market_cap_change_percentage_24h": -5.57949,
      "circulating_supply": 19065212,
      "total_supply": 21000000,
      "max_supply": 21000000,
      "ath": 93482,
      "ath_change_percentage": -57.95187,
      "ath_date": "2021-11-10T14:24:11.849Z",
      "atl": 72.61,
      "atl_change_percentage": 54035.65854,
      "atl_date": "2013-07-05T00:00:00.000Z",
      "roi": null,
      "last_updated": "2022-06-12T03:58:19.387Z"
      }
  ]})
    act(()=>render(<Trending />)) ;
    const loading= screen.getByTestId("loading")
    expect(loading).toBeInTheDocument();
    const coinContent = await screen.findByText('Bitcoin')
    expect(coinContent).toBeInTheDocument();
    const buttonEl = screen.queryByTestId("coinCircle")
    fireEvent.click(buttonEl)
    expect(mockedUsedNavigate).toHaveBeenCalledWith(`/coins/bitcoin`)
});

  
test('matches snapshot', () =>{
    const tree = renderer.create(<Trending/>).toJSON()
    expect(tree).toMatchSnapshot();
  })

