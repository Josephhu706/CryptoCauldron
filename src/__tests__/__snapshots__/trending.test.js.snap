// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  style={Object {}}
>
  <div
    data-testid="trendingPage"
  >
    <h4
      className="MuiTypography-root MuiTypography-h4 trendingTitle css-1s9j31j-MuiTypography-root"
      data-testid="title"
      style={
        Object {
          "fontFamily": "Montserrat",
          "margin": 18,
        }
      }
    >
      Top 10 Trending Coins By Volume
    </h4>
    <div
      className="app__header app__flex"
      data-testid="TrendingPage"
    >
      <div
        className="circlesContainer"
      >
        <div
          className="app__header-circlesTop3"
          style={
            Object {
              "opacity": 0,
              "transform": "translateY(300px) translateZ(0)",
            }
          }
        >
          <div
            className="circleContent"
          />
        </div>
        <div
          className="app__header-circlesTop7"
          style={
            Object {
              "opacity": 0,
              "transform": "translateY(300px) translateZ(0)",
            }
          }
        >
          <div
            className="circleContent"
          />
        </div>
        <div
          className="app__header-circlesTop10"
          style={
            Object {
              "opacity": 0,
              "transform": "translateY(300px) translateZ(0)",
            }
          }
        >
          <div
            className="circleContent"
          />
        </div>
      </div>
      <div
        data-testid="animation"
      >
        <div
          className="animation"
          style={
            Object {
              "height": 400,
              "width": 720,
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;
